File: Backtracking/LetterCombination/let.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    void doRecursion(vector<string>& ans, string digit, string helper, int idx){
        int n = digit.size();
        if(idx==n){
            ans.push_back(helper);
            return;
        }
        int offset = (digit[idx] - '2') * 3;
        if (digit[idx] > '7')
            offset++;
        int limit = (digit[idx] == '7' || digit[idx] == '9') ? 4 : 3;
        for(int i=0;i<limit;i++){
            char character = 'a' + offset + i;
            doRecursion(ans,digit,helper+character,idx+1);
        }
    }
    vector<string> letterCombinations(string digits) {
        vector<string>ans;
        doRecursion(ans,digits,"",0);
        return ans;
    }
};

File: Backtracking/LetterCombination/let.py
class Solution:
    def doRecursion(self, ans, arr, key, index):
        n = len(arr)
        if index == n:
            ans.append(key)
            return
        if arr[index] <= 1:
            self.doRecursion(ans, arr, key, index + 1)
            return
        offset = (arr[index]-2)*3
        if arr[index]>7: offset += 1
        limit = 3
        if arr[index]==7 or arr[index]==9: limit += 1
        for i in range(limit):
            character = chr(ord('a') + offset + i)
            self.doRecursion(ans,arr,key+character,index+1)
            
    def possibleWords(self, arr: list[int]) -> list[str]:
        # code here
        ans = []
        self.doRecursion(ans, arr, "", 0)
        return ans
